### DISCUSSION TITLE
Get waypoint's from players

### DISCUSSION THREAD

**Discussion by Tayipc38** [YimMenu/YimMenu#1502] [Ideas]
I have no idea how the game transfers waypoints from players in the same vehicle (it is just two floats at the end of the day). But this feature could be used to remotely fetch waypoints from players no? Ill be investigating but if anyone knows whether this event is done via the waypoint setter or the receiver, latter of which would make it easier, feel free to inform me how it is done.

`SET_APPLY_WAYPOINT_OF_PLAYER` is only used twice in the same function within freemode relating to CEO or MC president.
`NETWORK_GET_PLAYER_OWNS_WAYPOINT` Checks a bitfield within the player's net_object and I fear this native is unrelated to the waypoint transfer.

Logging script events and entering a players vehicle of which is certain they have a waypoint doesn't seem to send any specific events.

EDIT:

Getting waypoints can be done by hooking `48 89 5C 24 10 48 89 74 24 18 57 48 83 EC 30 8A`. CPed of the sender is located at `*(a1 - 1040)` and the waypoint data at `(a2 + 192)` and `(a2 + 196)` respectively. Thanks to @maybegreat48.

**Comment by maybegreat48** [comment]
From the 360 dump
```c++
NodeCommonDataOperations<CPlayerWaypointDataNode,IPlayerNodeDataAccessor>::`vftable`
    /* WARNING: Variable defined which should be unmapped: local_18 */
    /* WARNING: Variable defined which should be unmapped: local_10 */

    void FUN_82ff3cc8(int param_1,int *param_2)

    {
      char cVar1;
      undefined8 uVar2;
      double local_20;
      qword local_18;
      qword local_10;
      dword local_8;

      (**(code **)(*param_2 + 0x1c))(param_2,param_1 + 0x74,"Has Active Waypoint");
      if ((*(char *)(param_1 + 0x74) == '\0') &&
         (cVar1 = (**(code **)(*param_2 + 0x60))(param_2), cVar1 == '\0')) {
        *(undefined *)(param_1 + 0x75) = 0;
      }
      else {
        (**(code **)(*param_2 + 0x1c))(param_2,param_1 + 0x75,"Owns Waypoint");
        uVar2 = 0x40cf400000000000;
        (**(code **)(*param_2 + 0x3c))
                  (0x40cf400000000000,param_2,param_1 + 0x6c,*param_2,0x13,"Waypoint X");
        (**(code **)(*param_2 + 0x3c))(uVar2,param_2,param_1 + 0x70);
      }
      return;
    }
```
That node doesn't exist anymore, suggesting that it might have been merged with something else or had its name changed. `48 89 5C 24 10 48 89 74 24 18 57 48 83 EC 30 8A` might be it

**Reply by maybegreat48** [reply]
Function decomp
```c++
// CPlayerExtendedGameStateNode?
__int64 CPlayerUnkDataNode::Apply(__int64 plyr_interface, __int64 node_data)
{
  __int64 player; // rsi
  unsigned int v5; // eax
  int v6; // xmm1_4
  __int16 v7; // dx
  unsigned int v8; // eax
  bool v9; // zf
  float *v10; // rcx
  float v11; // xmm1_4
  __int64 result; // rax
  int v13; // [rsp+40h] [rbp+8h] BYREF
  int v14; // [rsp+44h] [rbp+Ch]

  player = *(plyr_interface - 1040);
  *(plyr_interface + 84) ^= (*(plyr_interface + 84) ^ (2 * *(node_data + 203))) & 2;
  *(plyr_interface + 84) ^= (*(plyr_interface + 84) ^ *(node_data + 202)) & 1;
  v5 = *(node_data + 204);
  if ( v5
    && v5 > *(plyr_interface + 80)
    && CNetGamePlayer::OwnsWaypoint(player)
    && CNetGamePlayer::CanSetWaypoint(player) ) // checks if we are in their vehicle seat
  {
    if ( *(node_data + 202) )                   // has waypoint data?
    {
      if ( *(node_data + 203) )                 // is waypoint set?
      {
        v6 = *(node_data + 196);
        v7 = *(node_data + 200);
        v13 = *(node_data + 192);
        v14 = v6;
        if ( !sub_7FF7E066BF8C(&v13, v7) )
        {
          DeleteWaypointsFromThisPlayer(0);
          v8 = sub_7FF7E13F9D68();
          CRadar::SetNewWaypointImpl(*(node_data + 192), *(node_data + 196), *(node_data + 200), 0, 0, v8);
        }
      }
    }
    else
    {
      DeleteWaypointsFromThisPlayer(0);
    }
  }
  v9 = *(plyr_interface - 1045) == 0;
  *(plyr_interface + 80) = *(node_data + 204);
  *(plyr_interface + 88) = *(node_data + 192);
  *(plyr_interface + 92) = *(node_data + 196);
  *(plyr_interface + 96) = *(node_data + 200);
  if ( v9 )
    v10 = (qword_7FF7E2470210 + 16);            // gets something from viewport
  else
    v10 = *(plyr_interface + 0xD0);
  v11 = *(node_data + 212) * 70.0;
  if ( v11 >= 7.5 )
  {
    if ( v11 > 70.0 )
      v11 = 70.0;
  }
  else
  {
    v11 = 7.5;
  }
  sub_7FF7E16FC42C(v10, v11, *(node_data + 216), v10[274], COERCE_INT(1000.0 - (*(node_data + 212) * 0.0)));// does some interesting math
  *(plyr_interface + 1216) = *(node_data + 208);
  *(plyr_interface + 1224) = *(node_data + 224);
  result = *(node_data + 220);
  *(*(player + 4264) + 3380i64) = result;
  return result;
}
```

**Reply by Tayipc38** [reply]
So this would be an incoming sync from the player that owns the waypoint and is prioritized?

**Reply by SkiddyToast** [reply]
It may not be from the player that owns the waypoint, for example they may be in a vehicle with another player and are using their waypoint. You’d need to check another bool that’s in that node to ensure the player actually owns it.

**Reply by Tayipc38** [reply]
Does the waypoint not remain after exiting the vehicle even though the waypoint wasn't original? So in that sense I would not care whether the waypoint is the players waypoint. If someone has a waypoint at all, I'd like to fetch it.

**Reply by maybegreat48** [reply]
You should try to figure out where the player interface is stored (most likely somewhere in `CNetObjPlayer`) or just hook the apply function and use `*(plyr_interface - 1040)` to find the owner. `(node_data + 192)` and `(node_data + 196)` are floats for the `x` and `y` coordinates of the waypoint. Some P2Cs can teleport players to their waypoints, so fetching them is definitely possible

**Reply by Tayipc38** [reply]
Without finding the plyr_interface I would have to rely on the apply function to execute before I get any info?

**Reply by maybegreat48** [reply]
Yep

**Reply by Tayipc38** [reply]
Little update. The player is defined by their CPed which is indeed located in plyr_interface - 1040. For anyone wondering this is a way you could fetch waypoints from players.
